mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -0,0 +1,547 @@
+  setDefault(fitIntercept -> true)
+
+  /**
+   * Set the maximum number of iterations.
+   * Default is 100.
+   * @group setParam
+   */
+  @Since("2.0.0")
+  def setMaxIter(value: Int): this.type = set(maxIter, value)
+  setDefault(maxIter -> 100)
+
+  /**
+   * Set the convergence tolerance of iterations.
+   * Smaller value will lead to higher accuracy with the cost of more iterations.
+   * Default is 1E-6.
+   * @group setParam
+   */
+  @Since("2.0.0")
+  def setTol(value: Double): this.type = set(tol, value)
+  setDefault(tol -> 1E-6)
+
+  /**
+   * Set the regularization parameter.
+   * Default is 0.0.
+   * @group setParam
+   */
+  @Since("2.0.0")
+  def setRegParam(value: Double): this.type = set(regParam, value)
+  setDefault(regParam -> 0.0)
+
+  /**
+   * Whether to over-/under-sample training instances according to the given weights in weightCol.
+   * If empty, all instances are treated equally (weight 1.0).
+   * Default is empty, so all instances have weight one.
+   * @group setParam
+   */
+  @Since("2.0.0")
+  def setWeightCol(value: String): this.type = set(weightCol, value)
+  setDefault(weightCol -> "")
+
+  /**
+   * Set the solver algorithm used for optimization.
+   * Currently only support "irls" which is also the default solver.
+   * @group setParam
+   */
+  @Since("2.0.0")
+  def setSolver(value: String): this.type = set(solver, value)
+  setDefault(solver -> "irls")
+
+  override protected def train(dataset: DataFrame): GeneralizedLinearRegressionModel = {
+    val familyLink = $(family) match {
+      case "gaussian" => if (isDefined(link)) Gaussian($(link)) else Gaussian("identity")
+      case "binomial" => if (isDefined(link)) Binomial($(link)) else Binomial("logit")
+      case "poisson" => if (isDefined(link)) Poisson($(link)) else Poisson("log")
+      case "gamma" => if (isDefined(link)) Gamma($(link)) else Gamma("inverse")
+    }
+
+    val numFeatures = dataset.select(col($(featuresCol))).limit(1).map {
+      case Row(features: Vector) => features.size
+    }.first()
+    if (numFeatures > 4096) {
+      val msg = "Currently, GeneralizedLinearRegression only supports number of features" +
+        s" <= 4096. Found $numFeatures in the input dataset."
+      throw new SparkException(msg)
+    }
+
+    val w = if ($(weightCol).isEmpty) lit(1.0) else col($(weightCol))
+    val instances: RDD[Instance] = dataset.select(
+      col($(labelCol)), w, col($(featuresCol))).map {
+      case Row(label: Double, weight: Double, features: Vector) =>
+        Instance(label, weight, features)
+    }
+
+    if ($(family) == "gaussian" && $(link) == "identity") {
+      val optimizer = new WeightedLeastSquares($(fitIntercept), $(regParam),
+        standardizeFeatures = true, standardizeLabel = true)
+      val wlsModel = optimizer.fit(instances)
+      val model = copyValues(new GeneralizedLinearRegressionModel(uid,
+        wlsModel.coefficients, wlsModel.intercept).setParent(this))
```

Comment:
better alignment, e.g.:

~~~scala
val model = copyValues(
  new GeneralizedLinearRegressionModel(uid, wlsModel.coefficients, wlsModel.intercept)
    .setParent(this))
~~~

The problem is `setParent` is in the same line as partial arguments from the constructors.